File: ios/WebyarNative/Sources/DesignSystem/Components/SendButton.swift
import SwiftUI

/// The one send button in the app.
///
/// There were four, and no two were alike: the chat composer's was 34pt with
/// a 17pt bold arrow, the voice-note bar's 38pt with a 16pt semibold one, the
/// internal-note sheet's 34pt with 16pt, and the mail composer's 34pt with
/// 15pt. Three of them showed nothing while the send was in flight. An
/// operator moving between an email and a chat saw the same control change
/// size and weight under their thumb.
///
/// One size, one weight, one disabled treatment, one busy treatment.
///
/// The three states are deliberately distinguishable without colour, because
/// the disabled and enabled fills differ only in opacity: the glyph swaps to
/// a spinner while sending, and the button stops taking taps in both of the
/// other two — so nothing here depends on telling two blues apart.
struct SendButton: View {
    /// Whether there is anything to send. Drives the fill and the tap.
    let isEnabled: Bool
    /// Whether a send is already in flight.
    var isSending: Bool = false
    /// What this button sends, spoken. Never drawn — the glyph is an arrow
    /// everywhere, so this is the only thing that says what it does.
    let label: String
    let action: () -> Void

    /// Matches the other controls that sit inside a composer pill, so the
    /// row's height is set by the pill rather than by whichever control is
    /// tallest today.
    private static let diameter: CGFloat = 34
    /// The box this button occupies in the row, and the box every other
    /// composer control occupies. The two differ so the circle has a point
    /// of air around it inside a box that still lines up with its neighbours.
    private static let box: CGFloat = ComposerGlyph.box

    var body: some View {
        Button(action: action) {
            ZStack {
                Circle()
                    .fill(Theme.Palette.brand.opacity(isEnabled && !isSending ? 1 : 0.35))

                if isSending {
                    ProgressView()
                        .controlSize(.small)
                        .tint(.white)
                } else {
                    Image(systemName: "arrow.up")
                        .font(.system(size: 15, weight: .bold))
                        .foregroundStyle(.white)
                }
            }
            .frame(width: Self.diameter, height: Self.diameter)
            .frame(width: Self.box, height: Self.box)
            // The tap target is Apple's 44, but it is given to the hit test
            // only. Laying it out at 44 made this the tallest thing in the
            // pill: the row bottom-aligns, so the arrow sat four points above
            // the glyphs beside it and the pill grew to hold a box nobody
            // could see. Pad out, take the hit test at the padded size, hand
            // the padding back to the layout.
            .padding((Theme.Size.minTouchTarget - Self.box) / 2)
            .contentShape(Rectangle())
            .padding(-(Theme.Size.minTouchTarget - Self.box) / 2)
        }
        .buttonStyle(SendButtonStyle())
        .disabled(!isEnabled || isSending)
        .accessibilityLabel(label)
        .accessibilityIdentifier(A11y.composerSend)
        .animation(Theme.Motion.standard, value: isEnabled)
        .animation(Theme.Motion.standard, value: isSending)
    }
}

/// Press feedback, and nothing else.
///
/// `.plain` leaves a send button completely inert under the thumb, which on
/// iOS reads as a tap that did not land — the reason people tap twice and
/// send twice. A small scale is what the system's own filled buttons do.
private struct SendButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.88 : 1)
            .animation(.spring(response: 0.22, dampingFraction: 0.7), value: configuration.isPressed)
    }
}

File: android/app/src/main/kotlin/com/webyar/operator/ui/components/SendButton.kt
package com.webyar.operator.ui.components

import androidx.compose.animation.animateColorAsState
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.material3.ripple
import androidx.compose.ui.semantics.Role
import com.webyar.operator.ui.design.Size
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.foundation.Canvas
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.draw.clip
import androidx.compose.foundation.background
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.webyar.operator.ui.A11y
import com.webyar.operator.ui.design.Motion

/**
 * The one send button in the app.
 *
 * There were four on iOS, and no two alike: 34pt with a 17pt bold arrow in
 * the chat composer, 38pt with 16pt semibold in the voice bar, 34/16 in the
 * note sheet, 34/15 in the mail composer. Three of them showed nothing while
 * a send was in flight. An operator moving between an email and a chat saw
 * the same control change size and weight under their thumb. One size, one
 * weight, one disabled treatment, one busy treatment.
 *
 * The three states are distinguishable without colour, deliberately: enabled
 * and disabled differ only in opacity, so the glyph swaps to a spinner while
 * sending and the button refuses taps in both of the other two. Nothing here
 * asks anyone to tell two blues apart.
 */
@Composable
fun SendButton(
    enabled: Boolean,
    /** What this sends, spoken. The glyph is an arrow everywhere, so this is
     *  the only thing that says what it does. */
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    sending: Boolean = false,
) {
    val interaction = remember { MutableInteractionSource() }
    val live = enabled && !sending
    val container by animateColorAsState(
        if (live) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.surfaceContainerHighest,
        Motion.effects(),
        label = "send-container",
    )
    val glyph by animateColorAsState(
        if (live) MaterialTheme.colorScheme.onPrimary else MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f),
        Motion.effects(),
        label = "send-glyph",
    )

    // A 48dp control — the touch target and the drawn circle are the same
    // size, Material 3 Expressive's medium icon button — whose circle
    // squares up under the thumb and springs back: the press is felt, so it
    // is not tapped twice, and sent twice.
    Box(
        modifier
            .size(Size.minTouchTarget)
            .clip(rememberPressShape(interaction, restPercent = 50f, pressedPercent = 28f))
            .background(container)
            .clickable(
                enabled = live,
                interactionSource = interaction,
                indication = ripple(color = MaterialTheme.colorScheme.onPrimary),
                role = Role.Button,
                onClick = onClick,
            )
            .testTag(A11y.COMPOSER_SEND)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        if (sending) {
            LoadingIndicator(color = MaterialTheme.colorScheme.primary, size = 30.dp)
        } else {
            UpArrow(color = glyph)
        }
    }
}

/**
 * The arrow, drawn rather than fetched.
 *
 * Up, not "send": it points out of the composer, and it does so in Persian
 * too — a paper-plane glyph is directional and would need mirroring, while an
 * arrow pointing up is the same arrow in every script.
 *
 * Drawn because `material-icons-core` is a curated subset and it is not worth
 * finding out at runtime whether a given name made the cut; the alternative,
 * `material-icons-extended`, is a very large artifact to pull in for one
 * glyph.
 */
@Composable
private fun UpArrow(color: Color, modifier: Modifier = Modifier) {
    Canvas(modifier.size(20.dp)) {
        val w = size.width
        val h = size.height
        val stroke = w * 0.145f
        val head = w * 0.30f
        val topY = h * 0.19f
        val midX = w / 2f

        drawLine(
            color = color,
            start = Offset(midX, h * 0.82f),
            end = Offset(midX, topY),
            strokeWidth = stroke,
            cap = StrokeCap.Round,
        )
        drawLine(
            color = color,
            start = Offset(midX - head, topY + head),
            end = Offset(midX, topY),
            strokeWidth = stroke,
            cap = StrokeCap.Round,
        )
        drawLine(
            color = color,
            start = Offset(midX + head, topY + head),
            end = Offset(midX, topY),
            strokeWidth = stroke,
            cap = StrokeCap.Round,
        )
    }
}
